File: typescript/ast/ast.ts
// Assuming you have a similar structure for `Token` in TypeScript
import { Token } from "../token/token";

export interface Node {
  tokenLiteral: string;
  string: string;
}

export interface Statement extends Node {
  // Empty in TypeScript, serves as a "marker" interface
}

export interface Expression extends Node {
  // Empty in TypeScript, serves as a "marker" interface
}

export class Program implements Node {
  statements: Statement[];

  constructor() {
    this.statements = [];
  }

  get tokenLiteral(): string {
    if (this.statements.length > 0) {
      return this.statements[0].tokenLiteral;
    } else {
      return "";
    }
  }

  get string(): string {
    return this.statements.map((s) => s.string).join("");
  }
}

export class LetStatement implements Statement {
  token: Token; // the token.LET token
  name: Identifier;
  value: Expression | null;

  constructor(token: Token, name: Identifier, value?: Expression) {
    this.token = token;
    this.name = name;
    this.value = value ?? null;
  }

  get tokenLiteral(): string {
    return this.token.literal;
  }

  get string() {
    return `${this.tokenLiteral} ${this.name.string} = ${
      this.value?.string ?? ""
    };`;
  }
}

export class Identifier implements Expression {
  token: Token; // the token.IDENT token
  value: string;

  constructor(token: Token, value: string) {
    this.token = token;
    this.value = value;
  }

  get tokenLiteral(): string {
    return this.token.literal;
  }

  get string() {
    return this.value;
  }
}

export class ReturnStatement implements Statement {
  token: Token; // the 'return' statement
  returnValue: Expression | null;

  constructor(token: Token, returnValue?: Expression) {
    this.token = token;
    this.returnValue = returnValue ?? null;
  }
  get tokenLiteral(): string {
    return this.token.literal;
  }

  get string() {
    return `${this.tokenLiteral} ${this.returnValue?.string ?? ""};`;
  }
}

export class ExpressionStatement implements Statement {
  token: Token;
  expression: Expression;

  constructor(token: Token, expression: Expression) {
    this.token = token;
    this.expression = expression;
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string() {
    return `${this.expression.string}`;
  }
}

export class BlockStatement implements Statement {
  token: Token;
  statements: Statement[];

  constructor(token: Token, statements: Statement[]) {
    this.token = token;
    this.statements = statements;
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string(): string {
    return this.statements.map((s) => s.string).join("");
  }
}

export class IntegerLiteral implements Expression {
  token: Token;
  value: number;

  constructor(token: Token, value: number) {
    this.token = token;
    this.value = value;
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string() {
    return this.token.literal;
  }
}

export class StringLiteral implements Expression {
  token: Token;
  value: string;

  constructor(token: Token, value: string) {
    this.token = token;
    this.value = value;
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string() {
    return this.token.literal;
  }
}

export class BooleanLiteral implements Expression {
  token: Token;
  value: boolean;

  constructor(token: Token, value: boolean) {
    this.token = token;
    this.value = value;
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string() {
    return this.token.literal;
  }
}

export class PrefixExpression implements Expression {
  token: Token;
  operator: string;
  right: Expression;

  constructor(token: Token, operator: string, right: Expression) {
    this.token = token;
    this.operator = operator;
    this.right = right;
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string() {
    return `(${this.operator}${this.right.string})`;
  }
}

export class InfixExpression implements Expression {
  token: Token;
  left: Expression;
  operator: string;
  right: Expression;

  constructor(
    token: Token,
    left: Expression,
    operator: string,
    right: Expression,
  ) {
    this.token = token;
    this.left = left;
    this.operator = operator;
    this.right = right;
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string() {
    return `(${this.left.string} ${this.operator} ${this.right.string})`;
  }
}

export class IfExpression implements Expression {
  token: Token;
  condition: Expression | null;
  consequence: BlockStatement | null;
  alternative: BlockStatement | null;

  constructor(
    token: Token,
    condition?: Expression,
    consequence?: BlockStatement,
    alternative?: BlockStatement,
  ) {
    this.token = token;
    this.condition = condition ?? null;
    this.consequence = consequence ?? null;
    this.alternative = alternative ?? null;
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string() {
    return `if (${this.condition?.string})${this.consequence?.string}`;
  }
}

export class FunctionLiteral implements Expression {
  token: Token;
  parameters: Identifier[];
  body: BlockStatement | null;
  name: string;

  constructor(token: Token) {
    this.token = token;
    this.parameters = [];
    this.body = null;
    this.name = "";
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string() {
    return `${this.tokenLiteral}${
      this.name ? this.name : ""
    }(${this.parameters.join(", ")}) ${this.body?.string}`;
  }
}

export class CallExpression implements Expression {
  token: Token;
  fn: Expression;
  arguments: Expression[];

  constructor(token: Token, fn: Expression) {
    this.token = token;
    this.fn = fn;
    this.arguments = [];
  }

  get tokenLiteral() {
    return this.token.literal;
  }

  get string() {
    return `${this.fn.string}(${this.arguments
      .map((a) => a.string)
      .join(", ")})`;
  }
}

File: python/monkey_ast/ast.py
from abc import ABC, abstractmethod
from typing import List
from typing_extensions import Optional
from monkey_token.token import Token


class Node(ABC):
    @abstractmethod
    def token_literal(self) -> str:
        pass

    @abstractmethod
    def string(self) -> str:
        pass


class Statement(Node, ABC):
    pass


class Expression(Node, ABC):
    pass


class Program(Node):
    def __init__(self):
        self.statements: List[Statement] = []

    def token_literal(self) -> str:
        if self.statements:
            return self.statements[0].token_literal()
        else:
            return ""

    def string(self):
        return "".join(map(lambda x: x.string(), self.statements))


class Identifier(Expression):
    def __init__(self, token: Token, value: str):
        self.token = token
        self.value = value

    def token_literal(self) -> str:
        return self.token.literal

    def string(self) -> str:
        return self.value


class LetStatement(Statement):
    def __init__(
        self,
        token: Token,
        name: Optional[Identifier] = None,
        value: Optional[Expression] = None,
    ):
        self.token = token
        self.name = name
        self.value = value

    def token_literal(self) -> str:
        return self.token.literal

    def string(self) -> str:
        parts = [self.token_literal() + " "]

        if self.name:
            parts.append(self.name.string() + " = ")

        if self.value:
            parts.append(self.value.string())

        parts.append(";")
        return "".join(parts)


class ReturnStatement(Statement):
    def __init__(self, token: Token, return_value: Optional[Expression] = None):
        self.token = token
        self.return_value = return_value

    def token_literal(self) -> str:
        return self.token.literal

    def string(self) -> str:
        parts = [self.token_literal() + " "]

        if self.return_value:
            parts.append(self.return_value.string())

        parts.append(";")
        return "".join(parts)


class ExpressionStatement(Statement):
    def __init__(self, token: Token, expression: Optional[Expression] = None):
        self.token = token
        self.expression = expression

    def token_literal(self) -> str:
        return self.token.literal

    def string(self) -> str:
        string = ""
        if self.expression:
            string = self.expression.string()
        return string


class IntegerLiteral(Expression):
    def __init__(self, token: Token, value: Optional[int] = None) -> None:
        self.token = token
        self.value = value

    def token_literal(self) -> str:
        return self.token.literal

    def string(self) -> str:
        return self.token.literal


class PrefixExpression(Expression):
    def __init__(self, token: Token, operator: str, right: Expression) -> None:
        self.token = token
        self.operator = operator
        self.right = right

    def token_literal(self) -> str:
        return self.token.literal

    def string(self) -> str:
        return f"({self.operator}{self.right.string()})"


class InfixExpression(Expression):
    def __init__(
        self, token: Token, left: Expression, operator: str, right: Expression
    ) -> None:
        self.token = token
        self.left = left
        self.operator = operator
        self.right = right

    def token_literal(self) -> str:
        return self.token.literal

    def string(self) -> str:
        return f"({self.left.string()} {self.operator} {self.right.string()})"


class BooleanLiteral(Expression):
    def __init__(self, token: Token, value: bool) -> None:
        self.token = token
        self.value = value

    def token_literal(self) -> str:
        return self.token.literal

    def string(self) -> str:
        return self.token.literal
